activerecord/lib/active_record/associations/belongs_to_association.rb

```diff
@@ -81,19 +81,25 @@ def saved_change_to_target?
 
       private
         def replace(record)
+          reset_inverse_instance(target)
```

Comment:
I think it would be better if we do this after the `if record` block.
In the case `raise_on_type_mismatch!(record)` raises the state won't be changed yet.

Reply:
Good catch 👍 !
If we move the `reset_inverse_instance(target)` after the `if record` block we'll loose the inverse when `target == record`. For this reason I've just  moved the change inside the `if record` and kept the existing `elsif` part.

Looks good to you?